[Driver][Frontend] Introduce `load-pass-plugin` option

antoniofrighetto opened the pull request:
Allow dynamic loading of LLVM passes via `load-pass-plugin` option passed to the Swift compiler driver, similarly to current Apple Clang `-fpass-plugin` option.

Previous discussion: https://forums.swift.org/t/load-external-llvm-pass-plugins-via-swift-frontend/67596

antoniofrighetto commented:
Please, let me know if anything needs to be improved as far as style conventions et alia is concerned, I'd be happy to do so.

antoniofrighetto commented:
@swift-ci please test

antoniofrighetto commented:
(cc/ @DougGregor)

antoniofrighetto commented:
Gentle ping. Any feedback or suggestions would be duly appreciated. Alongside this, I believe the option should be added as part of [Options.swift](https://github.com/apple/swift-driver/blob/main/Sources/SwiftOptions/Options.swift#L20) as well. May that be the case? (cc/ @DougGregor, @rintaro).

jslegendre commented:
Hey, this may not be getting much attention because the internal driver (this one) is considered legacy. The default driver as of Xcode 13 [swift-driver](https://github.com/apple/swift-driver). You might have better luck over there.

Note: I'm not affiliated with the Swift project in any way. Just a guy who'd love to see this merged

artemcm commented:
There was some discussion on this previously in:
https://github.com/apple/swift/pull/38017

The driver and frontend pieces are separate and yes, proper driver support would mean adding the logic in https://github.com/apple/swift-driver. Legacy (C++-based) driver support being optional/discouraged. 

@lhames @aschwaighofer can comment better on the architecture of loading such passes in IRGen.

antoniofrighetto commented:
Thanks a lot to both for the helpful insights. Support in the new swift-driver should be addressed by https://github.com/apple/swift-driver/pull/1472. The implementation should aim to ensure a clear separation between the driver and the frontend. Also, thanks for sharing #38017, was not aware of this. Reviewed the discussion there carefully, it was very informative.

antoniofrighetto commented:
Fixed conflicts by rebasing to main, and changed test library type from `MODULE` to `SHARED`, as CMake seems to be defaulting to generate the plugin with `.so` extension on macOS too (rather than `.dylib`, this was previously leading the unittest to fail).

weliveindetail commented on lib/IRGen/IRGen.cpp:
```diff
@@ -244,6 +261,18 @@ void swift::performLLVMOptimizations(const IRGenOptions &Opts,
+    } else {
+      diagnoseSync(Diags, DiagMutex, SourceLoc(),
+                   diag::unable_to_load_pass_plugin, PluginFile,
+                   toString(PassPlugin.takeError()));
```
I wonder if the diagnostic handling should remain in `performLLVM()`. Passing `Diags` and `DiagMutex` to `performLLVMOptimizations()` seems a bit unfortunate. We could instead just return the `llvm::Error` from here and let the callers take care of it. That would change the return type of the function. A lot of code in *include/swift* is using `llvm::Error` already, but I am not sure what the conventions are in Swift. @lhames What do you think?

weliveindetail reviewed (commented):
The way this patch proposes to load the plugin pass in IRGen looks good. It is exactly what we do in [Clang](https://github.com/llvm/llvm-project/blob/release/17.x/clang/lib/CodeGen/BackendUtil.cpp#L894-L903) and [Flang](https://github.com/llvm/llvm-project/blob/release/17.x/flang/lib/Frontend/FrontendActions.cpp#L885-L894), and what [the documentation](https://llvm.org/docs/WritingAnLLVMNewPMPass.html#registering-passes-as-plugins) recommends.

antoniofrighetto commented:
Gentle ping (cc/ @lhames).

wsmoses commented:
Just wanted to express support for this here as it also helps with adding plugins for AD

jslegendre commented:
Also expressing support for this!

antoniofrighetto commented:
Rebased to main, dropped unittest in `Driver` as part of recent legacy driver deprecation, added test coverage exercizing `performLLVMOptimization`. New driver support ready at https://github.com/apple/swift-driver/pull/1472. Kind ping @lhames, @aschwaighofer, @artemcm, @benlangmuir, @weliveindetail.

antoniofrighetto commented:
Kind ping (maybe cc/ @adrian-prantl).

adrian-prantl commented on lib/IRGen/IRGen.cpp:
```diff
@@ -194,7 +195,23 @@ static void align(llvm::Module *Module) {
+  Diags.diagnose(Loc, ID, std::move(Args)...);
+
+  if (DiagMutex)
+    DiagMutex->unlock();
```
`std::scoped<lock>(DiagMutex)`?

adrian-prantl commented on test/Frontend/Inputs/TestPlugin.cpp:
```diff
@@ -0,0 +1,50 @@
+//===---------------------- TestPlugin.cpp ----------------------*- C++ -*-===//
```
Nit: the `-*- C++ -*-` markers are only needed for `.h` files, because there it's ambiguous whether the language is C or C++.

adrian-prantl reviewed (approved):
LGTM with a couple of minor comments inside.

antoniofrighetto commented:
@adrian-prantl Addressed comments and rebased to main, thank you for reviewing this!

adrian-prantl commented:
@swift-ci test

antoniofrighetto commented:
Rebased to main, opened sister PR @ https://github.com/apple/llvm-project/pull/8678, reflecting changes (not sure if we should have targeted a different branch instead).

antoniofrighetto commented:
@adrian-prantl Kind ping on this, CI was previously failing.

antoniofrighetto commented:
Rebased to main (fixed conflicts), https://github.com/swiftlang/llvm-project/pull/8678 PR rebased to `stable/20230725`, which should be the expected branch. Very gentle remainder.

antoniofrighetto commented:
I assume this should be tested with:
```
Please test with following PR:
https://github.com/swiftlang/llvm-project/pull/8678

@swift-ci Please test
```

JDevlieghere commented:
https://github.com/swiftlang/llvm-project/pull/8678
https://github.com/swiftlang/swift-driver/pull/1472

@swift-ci test

JDevlieghere commented:
The macOS failure is:

```
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift/stdlib/private/SwiftPrivate/IO.swift:21:8: error: failed to build module 'Darwin'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.9.2 (swiftlang-5.9.2.2.11 clang-1500.1.0.2.2)', while this compiler is 'Apple Swift version 6.1-dev effective-5.10 (LLVM 97c65ef47934098, Swift c06508954544f05)'). Please select a toolchain which matches the SDK.
 19 | #else
 20 | #if canImport(Darwin)
 21 | import Darwin
    |        `- error: failed to build module 'Darwin'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.9.2 (swiftlang-5.9.2.2.11 clang-1500.1.0.2.2)', while this compiler is 'Apple Swift version 6.1-dev effective-5.10 (LLVM 97c65ef47934098, Swift c06508954544f05)'). Please select a toolchain which matches the SDK.
```

JDevlieghere commented:
https://github.com/swiftlang/llvm-project/pull/8678
https://github.com/swiftlang/swift-driver/pull/1472

@swift-ci test macos

artemcm commented:
@adrian-prantl @JDevlieghere is this good to merge?

adrian-prantl commented:
I don't have any further concerns with this patch.

antoniofrighetto committed:
[Driver][Frontend] Introduce `load-pass-plugin` option

Allow dynamic loading of LLVM passes via `load-pass-plugin`
option passed to the Swift compiler driver.

antoniofrighetto commented:
Rebased to main, fixed conflicts (to be tested again as above otherwise CI fails), kind ping for merging this.

aschwaighofer commented:
@swift-ci test

antoniofrighetto commented:
> @swift-ci test

Thanks, the other two PRs (https://github.com/swiftlang/llvm-project/pull/8678, https://github.com/swiftlang/swift-driver/pull/1472) are also required for the CI to pass.

aschwaighofer commented:
https://github.com/swiftlang/llvm-project/pull/8678
https://github.com/swiftlang/swift-driver/pull/1472
@swift-ci test

aschwaighofer commented:
@antoniofrighetto Can you put up a PR against `stable/20240723`. Swift has rebased against that newer llvm-project branch.

antoniofrighetto commented:
Done at https://github.com/swiftlang/llvm-project/pull/9586, thanks.

aschwaighofer commented:
https://github.com/swiftlang/swift-driver/pull/1472
https://github.com/swiftlang/llvm-project/pull/9586
@swift-ci test

antoniofrighetto commented:
Kind ping (before other rebases conflicts pop up :P)

al45tair commented:
@antoniofrighetto @aschwaighofer 
[The test for this is failing under ASAN](https://ci.swift.org/job/oss-swift-incremental-ASAN-RA-macos/7303/consoleFull#1161724776d6fdb6cb-f376-4f2e-8bce-d31c7304698b) because of an ODR rule violation (there are two vtables for `llvm::circular_raw_ostream`). See #77771.

paukala commented:
Great to see this merged!

Plugins (at least the one that I am using) usually rely on `libLLVM.dylib` to be linked with swiftc / clang, which seems to be an issue as `libLLVM.dylib` is not linked per default.

@antoniofrighetto did you experience something similar during development? If so, what would be a good way to include the needed lib?

In my quick test, I set `LLVM_BUILD_LLVM_DYLIB` and `LLVM_LINK_LLVM_DYLIB` to `true`. 

<details>
<summary>However, this resulted in `Library not loaded: @rpath/libLLVM.dylib` multiple errors while building the toolchain, such as:</summary>

 ```
dyld[43617]: Library not loaded: @rpath/libLLVM.dylib\n 
Referenced from: <...> /Users/user/swift-project/build/buildbot_osx/swift-macosx-arm64/bin/swift-frontend\n 
Reason: tried: \'/usr/lib/swift/libLLVM.dylib\' (no such file, not in dyld cache),
\'/System/Volumes/Preboot/Cryptexes/OS/usr/lib/swift/libLLVM.dylib\' (no such file),
\'/Users/user/swift-project/build/buildbot_osx/swift-macosx-arm64/lib/swift/host/compiler/libLLVM.dylib\' (no such file), 
\'/usr/lib/swift/libLLVM.dylib\' (no such file, not in dyld cache),
\'/System/Volumes/Preboot/Cryptexes/OS/usr/lib/swift/libLLVM.dylib\' (no such file),
\'/Users/user/swift-project/build/buildbot_osx/swift-macosx-arm64/lib/swift/host/compiler/libLLVM.dylib\' (no such file)\n")
```
</details>
